node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+        // tasks abort handles
+        let mut tracker = Tracker::default();
+        loop {
+            {
+                let message = {
+                    let mut ctx = ctx.clone();
+                    ctx.recv().await?
+                };
+                match message {
+                    FromOverseer::Communication { msg } => {
+                        // we signed this bitfield
+                        match msg {
+                            // Distribute a bitfield via gossip to other validators.
+                            BitfieldDistributionMessage::DistributeBitfield(
+                                hash,
+                                signed_availability,
+                            ) => {
+                                let msg = BitfieldGossipMessage {
+                                    relay_parent: hash,
+                                    signed_availability,
+                                };
+                                // @todo do we also make sure to not send it twice if we are the source
+                                // @todo this also apply to ourself?
+                                distribute(ctx.clone(), &mut tracker, msg).await?;
+                            }
+                            BitfieldDistributionMessage::NetworkBridgeUpdate(event) => {
+                                handle_network_msg(ctx.clone(), &mut tracker, event).await;
+                            }
+                        }
+                    }
+                    FromOverseer::Signal(OverseerSignal::StartWork(relay_parent)) => {
+                        let (validator_set, signing_context) =
+                            query_basics(ctx.clone(), relay_parent).await?;
+
+                        let _ = tracker.per_relay_parent.insert(
+                            relay_parent,
+                            PerRelayParentData {
+                                signing_context,
+                                validator_set: validator_set,
+                                ..Default::default()
+                            },
+                        );
+                        // futurama.insert(relay_parent.clone(), Box::pin(future.map(|_| ())));
+                        // active_jobs.insert(relay_parent.clone(), abort_handle);
+                    }
+                    FromOverseer::Signal(OverseerSignal::StopWork(relay_parent)) => {
+                        let _ = tracker.per_relay_parent.remove(&relay_parent);
+                    }
+                    FromOverseer::Signal(OverseerSignal::Conclude) => {
+                        tracker.per_relay_parent.clear();
+                        return Ok(());
+                    }
+                }
+            }
+            // active_jobs
+            //     .retain(|_, future| future.poll().is_pending());
+        }
+    }
+}
+
+/// modify the reputiation, good or bad
+async fn modify_reputiation<Context>(
+    mut ctx: Context,
```

Comment:
If the goal is to avoid borrowing issues across await points, then perhaps it would be better to give ownership of `Context` to some struct, and implement these functions as methods on that struct.